core/services/ocr2/plugins/promwrapper/plugin.go

```diff
@@ -80,93 +97,251 @@ var (
+}
+
+func (*defaultPometheusBackend) SetQueryToObservationLatency(labelValues []string, latency float64) {
+	promQueryToObservationLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPometheusBackend) SetObservationToReportLatency(labelValues []string, latency float64) {
+	promObservationToReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPometheusBackend) SetReportToAcceptFinalizedReportLatency(labelValues []string, latency float64) {
+	promReportToAcceptFinalizedReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPometheusBackend) SetAcceptFinalizedReportToTransmitAcceptedReportLatency(labelValues []string, latency float64) {
+	promAcceptFinalizedReportToTransmitAcceptedReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func New(
+	plugin types.ReportingPlugin,
+	name string,
+	chainType string,
+	chainID *big.Int,
+	config types.ReportingPluginConfig,
+	backend PrometheusBackend,
+) types.ReportingPlugin {
+	// Apply passed-in Prometheus backend if one is given.
+	var prometheusBackend PrometheusBackend = &defaultPometheusBackend{}
+	if backend != nil {
+		prometheusBackend = backend
+	}
+
 	return &promPlugin{
-		wrapped:      plugin,
-		name:         name,
-		chainType:    chainType,
-		chainID:      chainID,
-		oracleID:     fmt.Sprintf("%d", config.OracleID),
-		configDigest: common.Bytes2Hex(config.ConfigDigest[:]),
+		wrapped:                       plugin,
+		name:                          name,
+		chainType:                     chainType,
+		chainID:                       chainID,
+		oracleID:                      fmt.Sprintf("%d", config.OracleID),
+		configDigest:                  common.Bytes2Hex(config.ConfigDigest[:]),
+		queryEndTimes:                 make(map[types.ReportTimestamp]time.Time),
+		observationEndTimes:           make(map[types.ReportTimestamp]time.Time),
+		reportEndTimes:                make(map[types.ReportTimestamp]time.Time),
+		acceptFinalizedReportEndTimes: make(map[types.ReportTimestamp]time.Time),
+		prometheusBackend:             prometheusBackend,
 	}
 }
 
 func (p *promPlugin) Query(ctx context.Context, timestamp types.ReportTimestamp) (types.Query, error) {
 	start := time.Now().UTC()
 	defer func() {
 		duration := float64(time.Now().UTC().Sub(start))
-		promQuery.WithLabelValues(p.getLabelsValues()...).Observe(duration)
+		p.prometheusBackend.SetQueryDuration(getLabelsValues(p, timestamp), duration)
+		p.queryEndTimes[timestamp] = time.Now() // note time at end of Query()
 	}()
 
 	return p.wrapped.Query(ctx, timestamp)
 }
 
 func (p *promPlugin) Observation(ctx context.Context, timestamp types.ReportTimestamp, query types.Query) (types.Observation, error) {
 	start := time.Now().UTC()
+
+	// Report latency between Query() and Observation().
+	labelValues := getLabelsValues(p, timestamp)
+	if queryEndTime, ok := p.queryEndTimes[timestamp]; ok {
```

Comment:
Is `timestamp` here the same timestamp that is passed into `Query`?

Reply:
Let me dig more into OCR to answer this question better. My current understanding is that ReportTimestamp contains ConfigDigest, Epoch, and Round, so it will match for all the "Observation"s, "Report"s, and "Query"s etc.. inside the same round. It can then be assumed that Observe comes after Query and so on. However, this may be an assumption, so I can add an assertion that the rounds come correctly ordered.

Reply:
I think that assumption stands for Query, Observation, Report, but not for ShouldAcceptFinalizedReport and ShouldTransmitAcceptedReport. This comment from libocr:

```
// Don't make assumptions about the epoch/round order in which this function
// is called.
ShouldAcceptFinalizedReport(context.Context, ReportTimestamp, Report) (bool, error)
```

Reply:
Let me look at the code again though and see if your implementation handles this.

Reply:
We shouldn't get a ShouldAcceptFinalizedReport call with an incorrect timestamp anyway, and I don't think it matters that the order is undefined.

Reply:
You're right that `ShouldAcceptFinalizedReport` may be out-of-order in relation to other rounds, but probably not to other phases within the same round. And as you said, since we use a map lookup if they are out of order it will simply do nothing.